accepted/00063-hub-iss.md

```diff
@@ -0,0 +1,246 @@
+
+The current implementation has a code on the ISS Master to "export" files and code on the ISS Slave to "import" them - this concept is not supposed to change.
+
+### Database data import/export
+
+The ISS Master, given a specified piece of content to sync, should be able to export relevant data from the ISS Master database, to be transported and imported on the ISS Slave side.
+
+The current implementation serializes ISS Master database data to XML and deserializes it on the ISS Slave. Relatively complicated logic is in place on both communication ends to convert to and from this format.
+
+**First core idea** of this RFC is to **export database data from the ISS Master in form of a SQL script to be executed as-is on the ISS Slave** (eg. via `psql`).
+
+**Second core idea is to generate that SQL script from the ISS Master data mainly by inspecting the schema programmatically** (eg. figuring out by schema inspection most of the involved tables, key relationships, ordering, etc.).
+
+A proof-of-concept Python script that implements partially those ideas is available at: https://github.com/SUSE/spacewalk/blob/Manager-4.0-iss2/iss2/iss2.py
+
+#### First core idea: export as SQL script
+
+The SQL script exported from the ISS Master will have to be idempotent. For example:
+
+```sql
+INSERT INTO "rhnchecksumtype" (
+    "id",
+    "label",
+    "description"
+)
+VALUES
+    (1, 'md5', 'MD5sum'),
+    (2, 'sha1', 'SHA1sum'),
+    (3, 'sha256', 'SHA256sum'),
+    (4, 'sha384', 'SHA384sum'),
+    (5, 'sha512', 'SHA512sum')
+ON CONFLICT ("label") DO UPDATE
+    SET "description" = excluded."description";
+```
+
+The `ON CONFLICT DO UPDATE` clause is called an "upsert" effectively turning the `INSERT` into an `UPDATE` if a row with a conflict on the specified column (`label` in this case) is found.
+
+Note that additional mechanisms are needed to add correct values in the `id` column, if they come from a sequence. For example:
+
+```sql
+INSERT INTO "web_customer" (
+    "id",
+    "name",
+    "modified"
+)
+VALUES
+    (nextval('web_customer_id_seq'), 'SUS', '2019-10-17T10:09:11.750508+02:00'::timestamptz)
+ON CONFLICT ("name") DO UPDATE
+    SET "modified" = excluded."modified";
+```
+
+For tables with foreign keys, `SELECT` statements on parent tables will have to be used to link rows correctly, for example:
+
+```sql
+INSERT INTO "rhnchannelarch" (
+    "id",
+    "label",
+    "arch_type_id",
+    "name"
+)
+VALUES
+    (500, 'channel-ia32', (SELECT id FROM rhnarchtype WHERE label =  'rpm'), 'IA-32' - ),
+    (501, 'channel-ia32-deb', (SELECT id FROM rhnarchtype WHERE label =  'deb'), 'IA-32 Debian' - )
+ON CONFLICT ...
+```
+
+At this point it is believed/assumed that:
+  - ISS is by and large an additive process, so cleanups should not be needed for most tables. For those where it is necessary (eg. linking tables in an m:n relationship), `DELETE` statements either before or after `INSERT`s suffice
```

Comment:
When customers do cleanups, they often remove channels, and software packages and Errata. If we start to sync also images and containers I think they will be changed and removed much more often.

I think we should be sure that "Cleanup" is also working. I think this is important and will happen more often then we think.

Reply:
As per our Mumble chat: here I am referring to low-level, database row cleanups and not high-level, "entity" cleanups like removing an entire channel or an image.

The latter are not addressed by the current ISS but are always possible through regular UI/APIs, I expect that not to change with the proposal in this RFC.

I am now marking this resolved